Issue building OIIO 1.5, SIMD related error

Hi,

I am trying to build the current master branch and I am getting the following error:

In cmake-gui if I leave USE_SIMD empty, compiling fails and I get the following error message:
'''
...
[ 11%] Building CXX object src/libutil/CMakeFiles/simd_test.dir/simd_test.cpp.o
In file included from /root/oiio-customfix/src/libutil/simd_test.cpp:32:
/root/oiio-customfix/src/include/OpenImageIO/simd.h: In function ‘OpenImageIO::v1_5::simd::mask4 OpenImageIO::v1_5::simd::operator|(OpenImageIO::v1_5::simd::mask4, OpenImageIO::v1_5::simd::mask4)’:
/root/oiio-customfix/src/include/OpenImageIO/simd.h:333: error: cannot convert ‘float **vector**’ to ‘long long int **vector**’ for argument ‘1’ to ‘long long int **vector** _mm_or_si128(long long int __vector__, long long int **vector**)’
/root/oiio-customfix/src/include/OpenImageIO/simd.h: In function ‘const OpenImageIO::v1_5::simd::mask4 OpenImageIO::v1_5::simd::operator==(OpenImageIO::v1_5::simd::mask4, OpenImageIO::v1_5::simd::mask4)’:
/root/oiio-customfix/src/include/OpenImageIO/simd.h:348: error: cannot convert ‘OpenImageIO::v1_5::simd::mask4’ to ‘long long int **vector**’ for argument ‘1’ to ‘long long int **vector** _mm_cmpeq_epi32(long long int __vector__, long long int **vector**)’
make[2]: **\* [src/libutil/CMakeFiles/simd_test.dir/simd_test.cpp.o] Error 1
make[1]: **\* [src/libutil/CMakeFiles/simd_test.dir/all] Error 2
make: **\* [all] Error 2
'''
Then if I set USE_SIMD to 0, compiling fails too and I get the following error message:
'''
...
[ 11%] Building CXX object src/libutil/CMakeFiles/simd_test.dir/simd_test.cpp.o
In file included from /root/oiio-master2/src/libutil/simd_test.cpp:32:
/root/oiio-master2/src/include/OpenImageIO/simd.h:171: error: expected unqualified-id before ‘[’ token
/root/oiio-master2/src/include/OpenImageIO/simd.h:199: error: ISO C++ forbids declaration of ‘simd_t’ with no type
/root/oiio-master2/src/include/OpenImageIO/simd.h:199: error: expected ‘,’ or ‘...’ before ‘m’
/root/oiio-master2/src/include/OpenImageIO/simd.h:202: error: expected type-specifier before ‘simd_t’
/root/oiio-master2/src/include/OpenImageIO/simd.h:203: error: ‘simd_t’ does not name a type
/root/oiio-master2/src/include/OpenImageIO/simd.h:377: error: ‘simd_t’ does not name a type
/root/oiio-master2/src/include/OpenImageIO/simd.h: In constructor ‘OpenImageIO::v1_5::simd::mask4::mask4(int)’:
/root/oiio-master2/src/include/OpenImageIO/simd.h:199: error: class ‘OpenImageIO::v1_5::simd::mask4’ does not have any field named ‘m_vec’
/root/oiio-master2/src/include/OpenImageIO/simd.h:199: error: ‘m’ was not declared in this scope
/root/oiio-master2/src/include/OpenImageIO/simd.h: At global scope:
/root/oiio-master2/src/include/OpenImageIO/simd.h:450: error: expected unqualified-id before ‘[’ token
/root/oiio-master2/src/include/OpenImageIO/simd.h:484: error: expected ‘)’ before ‘m’
/root/oiio-master2/src/include/OpenImageIO/simd.h:487: error: expected type-specifier before ‘simd_t’
/root/oiio-master2/src/include/OpenImageIO/simd.h:488: error: ‘simd_t’ does not name a type
/root/oiio-master2/src/include/OpenImageIO/simd.h:806: error: ‘simd_t’ does not name a type
/root/oiio-master2/src/include/OpenImageIO/simd.h:922: error: expected unqualified-id before ‘[’ token
/root/oiio-master2/src/include/OpenImageIO/simd.h:971: error: ISO C++ forbids declaration of ‘simd_t’ with no type
/root/oiio-master2/src/include/OpenImageIO/simd.h:971: error: expected ‘,’ or ‘...’ before ‘m’
/root/oiio-master2/src/include/OpenImageIO/simd.h:974: error: expected type-specifier before ‘simd_t’
/root/oiio-master2/src/include/OpenImageIO/simd.h:975: error: ‘simd_t’ does not name a type
/root/oiio-master2/src/include/OpenImageIO/simd.h:1276: error: ‘simd_t’ does not name a type
/root/oiio-master2/src/include/OpenImageIO/simd.h: In constructor ‘OpenImageIO::v1_5::simd::float4::float4(int)’:
/root/oiio-master2/src/include/OpenImageIO/simd.h:971: error: class ‘OpenImageIO::v1_5::simd::float4’ does not have any field named ‘m_vec’
/root/oiio-master2/src/include/OpenImageIO/simd.h:971: error: ‘m’ was not declared in this scope
make[2]: **\* [src/libutil/CMakeFiles/simd_test.dir/simd_test.cpp.o] Error 1
make[1]: **\* [src/libutil/CMakeFiles/simd_test.dir/all] Error 2
make: **\* [all] Error 2
'''
Shouldn't that disable SSE completely?

Please note that OpenImageIO was compiling fine before the SIMD instructions directives implementation.
In any case, I'm wondering if I'm doing something wrong or if it's a bug.

Thanks in advance!

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -145,7 +145,7 @@ set (BUILD_OIIOUTIL_ONLY OFF CACHE BOOL "If ON, will build *only* libOpenImageIO
 set (USE_CPP11 OFF CACHE BOOL "Compile in C++11 mode")
 set (USE_LIBCPLUSPLUS OFF CACHE BOOL "Compile with clang libc++")
 set (EXTRA_CPP_ARGS "" CACHE STRING "Extra C++ command line definitions")
-set (USE_SIMD "" CACHE STRING "Use SIMD directives (0, sse2, sse3, sse4.1, sse4.2)")
+set (USE_SIMD "" CACHE STRING "Use SIMD directives (0, sse2, sse3, ssse3, sse4.1, sse4.2)")
 
 if (BUILDSTATIC AND ${CMAKE_SYSTEM_NAME} STREQUAL "Linux")
     # On Linux, the lack of -fPIC when building static libraries seems

--- a/Makefile
+++ b/Makefile
@@ -370,6 +370,6 @@ help:
 	@echo "      OIIO_BUILD_TESTS=0       Skip building the unit tests"
 	@echo "      BUILD_OIIOUTIL_ONLY=1    Build *only* libOpenImageIO_Util"
 	@echo "      USE_SIMD=arch            Build with SIMD support (choices: 0, sse2, sse3,"
-	@echo "                                    sse4.1, sse4.2)"
+	@echo "                                    ssse3, sse4.1, sse4.2)"
 	@echo ""
 

--- a/src/include/OpenImageIO/simd.h
+++ b/src/include/OpenImageIO/simd.h
@@ -371,7 +371,7 @@ public:
     /// Logical "or" component-by-component
     friend OIIO_FORCEINLINE mask4 operator| (mask4 a, mask4 b) {
 #if defined(OIIO_SIMD_SSE)
-        return _mm_or_si128 (a.m_vec, b.m_vec);
+        return _mm_or_ps (a.m_vec, b.m_vec);
 #else
         return mask4 (a.m_val[0] | b.m_val[0],
                       a.m_val[1] | b.m_val[1],
@@ -386,7 +386,7 @@ public:
     /// Equality comparison, component by component
     friend OIIO_FORCEINLINE const mask4 operator== (mask4 a, mask4 b) {
 #if defined(OIIO_SIMD_SSE)
-        return _mm_castsi128_ps (_mm_cmpeq_epi32 (a.m_vec, b.m_vec));
+        return _mm_castsi128_ps (_mm_cmpeq_epi32 (_mm_castps_si128 (a.m_vec), _mm_castps_si128(b.m_vec)));
 #else
         return mask4 (a[0] == b[0],
                       a[1] == b[1],
@@ -1033,10 +1033,10 @@ OIIO_FORCEINLINE int reduce_or (int4 v) {
 OIIO_FORCEINLINE int4 blend (int4 a, int4 b, mask4 mask)
 {
 #if defined(OIIO_SIMD_SSE) && OIIO_SIMD_SSE >= 4 /* SSE >= 4.1 */
-    return _mm_blendv_epi8 (a.simd(), b.simd(), mask);
+    return _mm_blendv_epi8 (a.simd(), b.simd(), _mm_castps_si128(mask));
 #elif defined(OIIO_SIMD_SSE) /* SSE2 */
-    return _mm_or_si128 (_mm_and_si128(mask, b.simd()),
-                         _mm_andnot_si128(mask.simd(), a.simd()));
+    return _mm_or_si128 (_mm_and_si128(_mm_castps_si128(mask.simd()), b.simd()),
+                         _mm_andnot_si128(_mm_castps_si128(mask.simd()), a.simd()));
 #else
     return int4 (mask[0] ? b[0] : a[0],
                  mask[1] ? b[1] : a[1],
@@ -1052,7 +1052,7 @@ OIIO_FORCEINLINE int4 blend (int4 a, int4 b, mask4 mask)
 OIIO_FORCEINLINE int4 blend0 (int4 a, mask4 mask)
 {
 #if defined(OIIO_SIMD_SSE)
-    return _mm_and_si128(mask, a.simd());
+    return _mm_and_si128(_mm_castps_si128(mask), a.simd());
 #else
     return int4 (mask[0] & a[0],
                  mask[1] & a[1],

